File: libnd4j/include/ops/declarable/generic/transforms/batch_to_space.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

//
// @author Yurii Shyrma (iuriish@yahoo.com)
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_batch_to_space)

#include <ops/declarable/headers/parity_ops.h>
#include <ops/declarable/helpers/s_t_b.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(batch_to_space, 2, 1, false, 0, 1) {
  // [bS*blockSize*blockSize, H/blockSize, W/blockSize, iC] is rearranged/permuted to [bS, oH, oW, iC]
  // oH = H - cropTop  - cropBottom
  // oW = W - cropLeft - cropRight

  auto input = INPUT_VARIABLE(0);
  auto crop = INPUT_VARIABLE(1);

  auto output = OUTPUT_VARIABLE(0);

  const LongType blockSize = INT_ARG(0);
  REQUIRE_TRUE(blockSize >= 2, 0, "BatchToSpace: integer parameter block_size must be >= 2, but got %i instead",
               blockSize);

  const int rank = input->rankOf();
  const int dim0 = input->sizeAt(0);
  REQUIRE_TRUE(rank == 4, 0, "BatchToSpace: rank of input array must be equal 4, but got %i instead", rank);
  REQUIRE_TRUE(dim0 % (blockSize * blockSize) == 0, 0,
               "BatchToSpace: first dimension of input array must be divisible by blockSize * blockSize (that is by "
               "%i), but got first dimension equal to %i",
               blockSize * blockSize, dim0);

  if (crop->sizeAt(0) != 2 || crop->sizeAt(1) != 2)
  REQUIRE_TRUE(false, 0, "BatchToSpace: operation expects crop shape to be {2, 2}, but got %s instead",
               ShapeUtils::shapeAsString(crop).c_str());

  const LongType cropBottom = crop->e<LongType>(0, 0);
  const LongType cropTop = crop->e<LongType>(0, 1);
  const LongType cropLeft = crop->e<LongType>(1, 0);
  const LongType cropRight = crop->e<LongType>(1, 1);

  const int oH = input->sizeAt(1) * blockSize - cropBottom - cropTop;  // top and bottom
  const int oW = input->sizeAt(2) * blockSize - cropLeft - cropRight;  // left and right
  REQUIRE_TRUE(oH >= 0, 0,
               "BatchToSpace: crop top/bottom values are too big and cause negative output height dimension !");
  REQUIRE_TRUE(oW >= 0, 0,
               "BatchToSpace: crop left/right values are too big and cause negative output width dimension !");

  if (shape::strideDescendingCAscendingF(input->shapeInfo()))
    helpers::batchToSpace(block.launchContext(), *input, *output, cropBottom, cropTop, cropLeft, cropRight, blockSize);
  else {
    auto dupped = input->dup(input->ordering());
    helpers::batchToSpace(block.launchContext(), *dupped, *output, cropBottom, cropTop, cropLeft, cropRight, blockSize);
    delete dupped;
  }
  return Status::OK;
}

////////////////////////////////////////////////////////////////////////////////
DECLARE_TYPES(batch_to_space) {
  getOpDescriptor()->setAllowedInputTypes(0, ANY)->setAllowedInputTypes(1, {ALL_INTS})->setSameMode(true);
}

////////////////////////////////////////////////////////////////////////////////
DECLARE_SHAPE_FN(batch_to_space) {
  auto inputShapeInfo = inputShape->at(0);
  auto cropShapeInfo = inputShape->at(1);

  const LongType blockSize = INT_ARG(0);
  REQUIRE_TRUE(blockSize >= 2, 0, "BatchToSpace: integer parameter block_size must be >= 2, but got %i instead",
               blockSize);

  const int rank = inputShapeInfo[0];
  const int dim0 = inputShapeInfo[1];
  REQUIRE_TRUE(rank == 4, 0, "BatchToSpace: rank of input array must be equal 4, but got %i instead", rank);
  REQUIRE_TRUE(dim0 % (blockSize * blockSize) == 0, 0,
               "BatchToSpace: first dimension of input array must be divisible by blockSize * blockSize (that is by "
               "%i), but got first dimension equal to %i",
               blockSize * blockSize, dim0);

  if (cropShapeInfo[1] != 2 || cropShapeInfo[2] != 2)
  REQUIRE_TRUE(false, 0, "BatchToSpace: operation expects crop shape to be {2, 2}, but got %s instead",
               ShapeUtils::shapeAsString(cropShapeInfo).c_str());

  const LongType cropBottom = INPUT_VARIABLE(1)->e<LongType>(0, 0);
  const LongType cropTop = INPUT_VARIABLE(1)->e<LongType>(0, 1);
  const LongType cropLeft = INPUT_VARIABLE(1)->e<LongType>(1, 0);
  const LongType cropRight = INPUT_VARIABLE(1)->e<LongType>(1, 1);

  const int oH = inputShapeInfo[2] * blockSize - cropTop - cropBottom;  // top and bottom
  const int oW = inputShapeInfo[3] * blockSize - cropLeft - cropRight;  // left and right
  REQUIRE_TRUE(oH >= 0, 0,
               "BatchToSpace: crop top/bottom values are too big and cause negative output height dimension !");
  REQUIRE_TRUE(oW >= 0, 0,
               "BatchToSpace: crop left/right values are too big and cause negative output width dimension !");

  // we always give out C order here
  return SHAPELIST(ConstantShapeHelper::getInstance().createShapeInfo(
      ArrayOptions::dataType(inputShapeInfo), 'c', {dim0 / (blockSize * blockSize), oH, oW, inputShapeInfo[4]}));
}

}  // namespace ops
}  // namespace sd

#endif

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/BatchToSpace.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;


import lombok.val;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;
import org.nd4j.linalg.factory.Nd4j;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * N-dimensional batch to space operation. Transforms data from a tensor from batch dimension into M spatial dimensions
 * according to the "blocks" specified (a vector of length M). Afterwards the spatial dimensions are optionally cropped,
 * as specified in "crops", a tensor of dim (M, 2), denoting the crop range.
 * <p>
 * Example:
 * input:        [[[[1]]], [[[2]]], [[[3]]], [[[4]]]]
 * input shape:  [4, 1, 1, 1]
 * blocks:       [2, 2]
 * crops:        [[0, 0], [0, 0]]
 * <p>
 * output:       [[[[1], [2]], [[3], [4]]]]
 * output shape: [1, 2, 2, 1]
 *
 * @author Max Pumperla
 */
public class BatchToSpace extends DynamicCustomOp {

    private int[] blocks;
    private int[][] crops;

    public BatchToSpace() {
    }

    public BatchToSpace(SameDiff sameDiff, SDVariable x, int[] blocks, int[] croppingTop, int... croppingBottom) {
        this(sameDiff, x, blocks, new int[][]{croppingTop, croppingBottom}, false);
    }

    public BatchToSpace(SameDiff sameDiff, SDVariable x, int[] blocks, int[][] crops, boolean inPlace) {
        this(sameDiff, new SDVariable[]{x}, blocks, crops, inPlace);
    }

    public BatchToSpace(SameDiff sameDiff, SDVariable[] args, int[] blocks, int[][] crops, boolean inPlace) {
        super(null, sameDiff, new SDVariable[]{args[0], sameDiff.constant(Nd4j.createFromArray(crops))}, inPlace);

        this.blocks = blocks;
        this.crops = crops;

        for (val b : blocks)
            addIArgument(b);
    }

    public BatchToSpace(INDArray x, int[] blocks, int[] croppingTop, int... croppingBottom) {
        addInputArgument(x);
        int[][] crops = new int[][]{croppingTop, croppingBottom};
        this.blocks = blocks;
        this.crops = crops;

        for (val b : blocks)
            addIArgument(b);
    }


    @Override
    public String opName() {
        return "batch_to_space";
    }

    @Override
    public String onnxName() {
        return "batch_to_space";
    }

    @Override
    public String tensorflowName() {
        return "BatchToSpace";
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> i_v) {
        // Inverse of batch to space is space to batch with same blocks and padding as crops
        SDVariable gradient = sameDiff.setupFunction(i_v.get(0));
        return Arrays.asList(sameDiff.cnn().spaceToBatch(gradient, blocks, crops[0], crops[1]));
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes){
        return Collections.singletonList(dataTypes.get(0));
    }
}
